### PR TITLE
Add support for User Defined Type as wide parameters

### PR DESCRIPTION
**Pull request by jfalcou**
Adapted the code base so any kumi::product_type can be used within wide.
Support for externally yet default provided ==, !=, stream insertion and if_else is done..

### CODE REVIEW

File: include/eve/arch/cpu/wide.hpp [hunk 1]
```diff
@@ -129,7 +132,7 @@ inline namespace EVE_ABI_NAMESPACE
     //! Constructs a eve::wide by splatting a scalar value in all lanes
     template<scalar_value S>
     EVE_FORCEINLINE explicit  wide(S const& v)  noexcept
-                            : storage_base(detail::make(eve::as<wide>{}, v))
+                            : storage_base(detail::make(eve::as<wide>{}, static_cast<Type>(v)))
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:135]: Oh. Hmm. Oh.

THis means we ignore all checks. Hmm. Not sure it's good.

File: include/eve/arch/cpu/wide.hpp [hunk 2]
```diff
@@ -138,7 +141,20 @@ inline namespace EVE_ABI_NAMESPACE
...
+                  : storage_base(detail::make ( eve::as<wide>{}
+                                              , static_cast<Type>(v0)
+                                              , static_cast<Type>(v1)
+                                              , static_cast<Type>(vs)...
+                                )             )
+    {}
+
+    //! Constructs a eve::wide from a sequence of SIMD product type values
+    template<simd_value S0, simd_value... Ss>
+    EVE_FORCEINLINE wide( S0 v0, Ss... vs) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:153]: explicit, since it can be of one parameter.

File: include/eve/arch/cpu/wide.hpp [hunk 3]
```diff
@@ -869,45 +885,73 @@ inline namespace EVE_ABI_NAMESPACE
...
+    friend EVE_FORCEINLINE auto operator==(S v, wide const& w) noexcept
     {
       return w == v;
     }
 
+    template<scalar_value S>
+    friend EVE_FORCEINLINE auto operator==(S v, wide const& w) noexcept
+            requires(detail::tag_dispatchable<tag::is_equal_,S,wide>)
+    {
+      return tagged_dispatch(tag::is_equal_{}, v, w);
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:918]: Why isn't `is_equal` deal with this?

File: include/eve/detail/function/simd/common/friends.hpp [hunk 4]
```diff
@@ -135,7 +142,24 @@ namespace eve::detail
...
   //================================================================================================
-  template<real_simd_value Wide>
+  template<simd_value Wide>
+  EVE_FORCEINLINE auto self_eq(Wide const& v,Wide const& w) noexcept
+  requires( kumi::product_type<element_type_t<Wide>> )
+  {
+    if constexpr( detail::tag_dispatchable<tag::is_equal_,Wide,Wide> )
+    {
+      return tagged_dispatch(tag::is_equal_{}, v, w);
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/friends.hpp:152]: This should just kick in, no?

File: include/eve/arch/cpu/wide.hpp [hunk 5]
```diff
@@ -869,45 +885,73 @@ inline namespace EVE_ABI_NAMESPACE
...
     //! @brief Element-wise equality comparison of a scalar value and a eve::wide
     template<scalar_value S>
-    friend EVE_FORCEINLINE logical<wide> operator==(S v, wide w) noexcept
+    friend EVE_FORCEINLINE auto operator==(S v, wide const& w) noexcept
     {
       return w == v;
     }
 
+    template<scalar_value S>
+    friend EVE_FORCEINLINE auto operator==(S v, wide const& w) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:915]: Do you want equal/not_equal to be enabled through tagged_dispatch?

I thought we'd go with:

```
struct point : kumi::tuple<int, int, int> {
   friend decltype(auto) x(eve::compatible<point> auto&& self) ...
   friend decltype(auto) y(eve::compatible<point> auto&& self) ...
   friend decltype(auto) z(eve::compatible<point> auto&& self) ...
   
   auto operator==(eve::compatible<point> auto const& a, eve::compatible<point> auto const& b) {
      return x(a) == x(b) && y(a) == y(b) && z(a) == z(b);
   } 
};
```

Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:915]: (whatever you want to call the concept)

File: include/eve/detail/function/simd/common/friends.hpp [hunk 6]
```diff
@@ -135,7 +142,24 @@ namespace eve::detail
...
+    else
+    {
+      auto zipped = kumi::zip(v.storage(),w.storage());
+      return kumi::fold_right ( [](auto acc , auto e) { return acc && ( get<0>(e) == get<1>(e) ); }
+                              , zipped, true
+                              );
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/friends.hpp:159]: Does this use wide equality?

File: include/eve/detail/function/simd/common/friends.hpp [hunk 7]
```diff
@@ -152,7 +176,24 @@ namespace eve::detail
...
+    {
+      return tagged_dispatch(tag::is_not_equal_{}, v, w);
+    }
+    else
+    {
+      auto zipped = kumi::zip(v.storage(),w.storage());
+      return kumi::fold_right ( [](auto acc , auto e) { return acc || ( get<0>(e) != get<1>(e) ); }
+                              , zipped, false
+                              );
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/friends.hpp:193]: I think we should pick up equal customization  for neq

File: include/eve/arch/cpu/wide.hpp (hunk 5 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:915]: As per the slack discussion, we want to limit the amount of APi for specialisation. I decided `tagged_dispatch` was a good vessel for that.

File: include/eve/arch/cpu/wide.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:918]: Simplified as per slack discussion

File: include/eve/detail/function/simd/common/friends.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/common/friends.hpp:152]: Simplified as per slack discussion

File: include/eve/arch/cpu/wide.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:153]: Done

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:135]: What do you mean ?

File: include/eve/detail/function/simd/common/friends.hpp (hunk 6 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/common/friends.hpp:159]: each member is compared using the == of the wide beneath yes

File: include/eve/detail/function/simd/common/friends.hpp (hunk 7 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/common/friends.hpp:192]: We now pick up the tagged_dispatch of is_equal if it is available. For the basic one, I think it's better to do piecewise !=.

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:135]: `static_cast<Type>` will force conversion from int to char for example.

File: include/eve/arch/cpu/wide.hpp [hunk 8]
```diff
@@ -869,41 +885,41 @@ inline namespace EVE_ABI_NAMESPACE
     // Logical operations
     //==============================================================================================
     //! @brief Element-wise equality comparison of two eve::wide
-    friend EVE_FORCEINLINE logical<wide> operator==(wide v, wide w) noexcept
+    friend  EVE_FORCEINLINE auto operator==(wide const& v, wide const& w) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:888]: Oh come on, we can pass by value. But w/e.

File: include/eve/detail/function/simd/common/friends.hpp (hunk 7 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/common/friends.hpp:192]: What about opt in for neq. We don't care?

Comment (jfalcou) [reply include/eve/detail/function/simd/common/friends.hpp:192]: it was removed

File: include/eve/arch/cpu/wide.hpp (hunk 8 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:888]: this is a remnant of an old test, will clean up

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**